connectors/sources/network_drive.py

```diff
@@ -419,49 +417,90 @@ async def traverse_diretory(self, path):
+        while stack:
+            directory_info = []
+            current_path = stack.pop()
+            try:
+                directory_info = await asyncio.to_thread(
+                    partial(
+                        smbclient.scandir,
+                        path=current_path,
+                        port=self.port,
+                    ),
+                )
+            except SMBConnectionClosed as exception:
+                self._logger.exception(
+                    f"Connection got closed. Error {exception}. Registering new session"
+                )
+                await asyncio.to_thread(self.smb_connection.create_connection)
+                raise
+            except (SMBOSError, SMBException) as exception:
+                self._logger.exception(
+                    f"Error while scanning the path {current_path}. Error {exception}"
+                )
+                continue
+
+            for file in directory_info:
+                if file.is_dir():
+                    stack.append(file.path)
+                file_path = file.path.split("/", 1)[1].replace("\\", "/")
+                is_file_match = glob.globmatch(
+                    file_path, glob_pattern, flags=glob.GLOBSTAR
+                )
+                if is_file_match:
+                    yield self.format_document(file=file)
+
+    async def fetch_filtered_directory(self, advanced_rules):
         """
-        Find matching paths based on advanced rules.
+        Fetch file and folder based on advanced rules.
 
         Args:
             advanced_rules (list): List of advanced rules configured
 
         Returns:
-            matched_paths (set): Set of paths that match the advanced rules.
-            invalid_rules (list): List of advanced rules that have no matching paths.
+            format_document: Formatted document based on advance rules
         """
         self._logger.debug(
-            "Fetching the matched directory paths using the list of advanced rules configured"
+            "Fetching the matched directory/files using the list of advanced rules configured"
         )
-        invalid_rules = []
-        matched_paths = set()
+        unmatched_rules = set()
         for rule in advanced_rules:
-            rule_valid = False
+            rule_matched = False
             glob_pattern = rule["pattern"].replace("\\", "/")
-            paths = await self.get_directory_details()
-            for path, _, _ in paths:
-                normalized_path = path.split("/", 1)[1].replace("\\", "/")
-                is_match = glob.globmatch(
-                    normalized_path, glob_pattern, flags=glob.GLOBSTAR
-                )
+            async for document in self.traverse_directory_for_syncrule(
+                path=rf"\\{self.server_ip}/{self.drive_path}",
+                glob_pattern=glob_pattern,
+            ):
+                yield document
+                rule_matched = True
```

Comment:
I don't think it works well when specifying multiple glob patterns - it will repeat traversal for each pattern. Can we do one traversal this way?

Reply:
yes, it will repeat the traversal for each pattern but if we do like this:
```
async for document in self.traverse_directory_for_syncrule(
                path=rf"\\{self.server_ip}/{self.drive_path}",
                glob_pattern=glob_pattern,
            ):
            for rule in rules:
                if match:
                    yield document
```
In this case there can be a time, when file/folder doesn't match with any of the rule and this can happen continuously for n number of documents. In that case connector will stop syncing until it finds the match.

Reply:
That's right, but I also see in code that you don't check if the folder you're traversing through matches the glob, e.g. imagine:

You have a glob "\something\**\*.txt" and you have a 20GB directory "\images" - why traverse it if it did not satisfy the glob already?

This optimisation should eliminate the problem and make things fast.